PYIC-6724: Implement inbuilt back navigation

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Implement inbuilt back navigation

### Why did it change

[PYIC-6724: Start reading user state from stack](https://github.com/govuk-one-login/ipv-core-back/commit/200ee830e3446da80e3d49e7d63677d89e0eee50)

This switches core-back to start using the users state and subjourney
type from the stack, rather than the individual fields in their session.
It removes those fields entirely.
This also updates the exception thrown if the stack is empty to a
runtime exception. Keeping it as checked would have made the
process-journey-event lambda excessivley verbose. And this is an
exception that shouldn't ever happen in reality.

[PYIC-6724: Implement back functionality in journey-engine](https://github.com/govuk-one-login/ipv-core-back/commit/39b7631d772da8030b491fa3d98d1ddea11ecd28)

This adds functionality to the journey engine to detect a back event,
and find a legitimate previous state for them from the stack.

Initially we will only support back where the current state and the
previous state are page states.

This change needs to take account of jumps between subjourneys, as well
as nested journeys.

This change removes the addition of JourneyChangeStates to the stack -
they make the logic a lot more complicated and add little of value.

This change also adds a new field to the states object - journeyType.
This in populated when the state machine is initialized, and makes
writing states to the stack a lot simpler, particularly when moving
between subjourneys.

When a user does go "back", this change actually adds the previous state
on top of the stack, rather than just popping off the top one. This will
give us a fuller picture of the users journey should we need it.

[PYIC-6724: Default to back event on state](https://github.com/govuk-one-login/ipv-core-back/commit/b4b00bf368f70b1068f39b94ba432b4d984e5745)

If a state has a back event explicitly defined we should use it rather
than the inbuilt mechanism. The motivation for this is that we currently
have back events defined that emit audit events defined in the journey
map. If we just use the inbuilt functionality we'd lose the audit
events.

This will also allow us to configure a back event that does something
other than go back to the user previous state, or to a non page state.
Such as in the passport mitigation journey.

[PYIC-6724: Remove back events from journey map](https://github.com/govuk-one-login/ipv-core-back/commit/e8d4a599b1fb79699abf9b32cbb30c4f3d92be2b)

This removes all the back events that the new functionality can replace,
leaving only those with audit events and those that go back to a
non-page state (CRI states).

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-6724](https://govukverify.atlassian.net/browse/PYIC-6724)


[PYIC-6724]: https://govukverify.atlassian.net/browse/PYIC-6724?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-6724: Start reading user state from stack

This switches core-back to start using the users state and subjourney
type from the stack, rather than the individual fields in their session.
It removes those fields entirely.
This also updates the exception thrown if the stack is empty to a
runtime exception. Keeping it as checked would have made the
process-journey-event lambda excessivley verbose. And this is an
exception that shouldn't ever happen in reality.

Wynndow committed:
PYIC-6724: Implement back functionality in journey-engine

This adds functionality to the journey engine to detect a back event,
and find a legitimate previous state for them from the stack.

Initially we will only support back where the current state and the
previous state are page states.

This change needs to take account of jumps between subjourneys, as well
as nested journeys.

This change removes the addition of JourneyChangeStates to the stack -
they make the logic a lot more complicated and add little of value.

This change also adds a new field to the states object - journeyType.
This in populated when the state machine is initialized, and makes
writing states to the stack a lot simpler, particularly when moving
between subjourneys.

When a user does go "back", this change actually adds the previous state
on top of the stack, rather than just popping off the top one. This will
give us a fuller picture of the users journey should we need it.

Wynndow committed:
PYIC-6724: Default to back event on state

If a state has a back event explicitly defined we should use it rather
than the inbuilt mechanism. The motivation for this is that we currently
have back events defined that emit audit events defined in the journey
map. If we just use the inbuilt functionality we'd lose the audit
events.

This will also allow us to configure a back event that does something
other than go back to the user previous state, or to a non page state.
Such as in the passport mitigation journey.

Wynndow committed:
PYIC-6724: Remove back events from journey map

This removes all the back events that the new functionality can replace,
leaving only those with audit events and those that go back to a
non-page state (CRI states).

Joe-Edwards-GDS commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/domain/JourneyState.java:
```diff
@@ -0,0 +1,5 @@
+package uk.gov.di.ipv.core.library.domain;
+
+public record JourneyState(IpvJourneyTypes subJourney, String state) {
+    public static final String JOURNEY_STATE_DELIMITER = "/";
```
If you're putting this here, maybe the serialisation/deserialisation should be too?

Wynndow replied:
Which (de)serialisation? That's being handled for us now.
Do you mean the convenience stuff in the ipvSessionItem? If so, I'll see what I can squeeze in here.

Joe-Edwards-GDS replied:
Yeah, I don't really mind - I think it's reasonable to leave serialisation as an implementation detail of the session, but then it's odd to keep the delimiter here!

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/ProcessJourneyEventHandler.java:
```diff
@@ -253,41 +252,51 @@ private StepResponse executeJourneyEvent(
+            var previousJourneyState = ipvSessionItem.getPreviousState();
+
+            if (isPageState(initialJourneyState) && isPageState(previousJourneyState)) {
+                ipvSessionItem.pushState(previousJourneyState);
```
Do you think this is right - will this mean if you keep pressing back you'll toggle back-and-forth, rather than continuing to go back?

(Should it be `popState()` instead?)

Wynndow replied:
I guess I was thinking that the user is only going to go 'back' once, and this would give us a fuller picture of there journey in the stack. But you're right, any more than once and it'll break. I'll switch it, it's probably not that important to have that information in the stack anyway.

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/ProcessJourneyEventHandler.java:
```diff
@@ -253,41 +252,51 @@ private StepResponse executeJourneyEvent(
-                                ipvSessionItem.getJourneyType().name())));
+                                initialJourneyState.subJourney())));
+
+        if (BACK_EVENT.equals(journeyEvent) && !isBackEventDefinedOnState(initialJourneyState)) {
```
Do we have cases where `back` goes to a page that is not just the previous state? That's a bit annoying - but I guess a handy feature to allow.

Wynndow replied:
I don't think so? If you're referring to the second condition here, that's to allow us to override the default back behaviour. Useful for where we now have back events with an audit event attached in the journey map, and we also have a couple of cases of a back event leading to a CRI state.

Joe-Edwards-GDS replied:
Yeah, the second condition. Going 'back' to a CRI state sounds odd - I wonder if they are really 'back' steps.

Still useful capability though :)

Wynndow replied:
Sort of is - it's for when a user is at a CRI and then abandon. We now show them a screen and let them go back to it if they want to try again - part of the F2F volumes stuff.

Although it is done from a radio button rather than a back link, so perhaps it should use an event other than back...

Joe-Edwards-GDS replied:
Ah ok - I might consider it more a 'try again' than a 'back', but don't worry too much for this PR.

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/ProcessJourneyEventHandler.java:
```diff
@@ -413,4 +417,33 @@ private void sendSubJourneyStartAuditEvent(
+                        stateMachines.get(journeyState.subJourney()).getState(journeyState.state()))
+                .getEvents()
+                .containsKey(BACK_EVENT);
+    }
```
It doesn't quite feel right to have all these at the handler level, but I guess the trouble is you can't easily move it inside `StateMachine` because it might move between them :/

There's probably a refactoring begging to be let out, but perhaps not the time. (I also remember we had that chat about trying to rewrite this all in TS - perhaps that's the real answer :D )

Wynndow replied:
Yeah I agree - I wanted to push as much of this into the state machine, but it's hard hecause of the way our collection of state machines is structured. It would perhaps be better to just have on state machine that knows about all the journey maps.

Joe-Edwards-GDS replied:
Yep, agreed - the distinction is really for ease of definition. Optimising the execution isn't really that important (not until we get _really_ complicated!)

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/ProcessJourneyEventHandler.java:
```diff
@@ -413,4 +417,33 @@ private void sendSubJourneyStartAuditEvent(
+        return stateMachine.isPageState(journeyState);
+    }
+
+    private BasicState journeyStateToBasicState(JourneyState journeyState) {
```
Does this need to return a basic state?

Wynndow replied:
I think it returns a basic staate because it always will, given the context it's used in - it's only used on a journeyState we've already checked is a BasicState.

Joe-Edwards-GDS replied:
I think I meant - does the caller actually care that this is a Basic State?

Wynndow replied:
I think not - will get it to return the interface

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/ProcessJourneyEventHandler.java:
```diff
@@ -413,4 +417,33 @@ private void sendSubJourneyStartAuditEvent(
+    }
+
+    private boolean isBackEventDefinedOnState(JourneyState journeyState) {
+        return ((BasicState)
```
Can we guarantee this is a basic state?

Wynndow replied:
No, possibly not. I'll update it to check before trying to cast it.

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/main/java/uk/gov/di/ipv/core/processjourneyevent/statemachine/StateMachine.java:
```diff
@@ -54,6 +59,36 @@ public TransitionResult transition(
+                            .getNestedJourneyDefinition()
+                            .getNestedJourneyStates(),
+                    stateParts);
+        }
```
A bit sad, but seems hard to avoid without a major refactor :/

The JJJ does it by exploding out the nested states during initialisation instead (so once the map is initialised there are no 'nested' states to deal with), which might be something to consider if we end up rewriting this.

Wynndow replied:
Yeah, that might be a good way to go. The nested stuff is too complex.

Wynndow replied:
Alternatively we just bin nested states. They're not as useful as we hoped they would be

Joe-Edwards-GDS replied:
Yeah, they might need overhauling slightly to get maximum value. They've been kinda useful for Address+Fraud, but it wouldn't have been that hard to just define them each time either.

Joe-Edwards-GDS commented on lambdas/process-journey-event/src/test/java/uk/gov/di/ipv/core/processjourneyevent/JourneyMapTest.java:
```diff
@@ -206,6 +206,9 @@ private void findPageSpecificStatesAndEvents(
 
                     var pageEvents = new HashSet<>(basicState.getEvents().keySet());
+                    pageEvents.remove(
+                            "back"); // the back event is a special case that may or may not be
```
Yeah - I also considered the option of still requiring it to be explicitly defined (but with no need to explicitly specify a target).

That would also let you attach audit events - which might be helpful for the user details aborted case?

```
events:
  next:
    targetState: something
  back:
  other:
    targetState: something


events:
  back:
    auditEvents:
      - IPV_GONE_BACK
```

Wynndow replied:
Yeah - feels like we're edging closer to what we already have though - meaning we'd always need to go consult the journey map anyway.

Joe-Edwards-GDS replied:
I guess we can add this if we need it

Wynndow committed:
PYIC-6724: Pop states for back functionality

This updates the way the back functionality works. Rather than appending
the user previous state to the head of the stack, this pops their
current state off. This allows for multiple 'back' events, rather than
getting the user stuck in a loop.

Wynndow committed:
PYIC-6724: Ensure state is basic before checking events

This adds a check to ensure that a state is a basic state before
checking if back is defined in its events. If its not a basic state we
just return false

Wynndow committed:
PYIC-6724: Move JourneyState serialisation

This moves the converting of a JourneyState to a string to the
JourneyState record.
